packages/priority-queue/src/index.js

```diff
@@ -67,61 +67,62 @@ import requestIdleCallback from './request-idle-callback';
  * @return {WPPriorityQueue} Queue object with `add`, `flush` and `reset` methods.
  */
 export const createQueue = () => {
-	/** @type {WPPriorityQueueContext[]} */
-	let waitingList = [];
-
-	/** @type {WeakMap<WPPriorityQueueContext,WPPriorityQueueCallback>} */
-	let elementsMap = new WeakMap();
-
+	/** @type {Map<WPPriorityQueueContext, WPPriorityQueueCallback>} */
+	const waitingList = new Map();
 	let isRunning = false;
 
 	/**
 	 * Callback to process as much queue as time permits.
 	 *
+	 * Map Iteration follows the original insertion order. This means that here
+	 * we can iterate the queue and know that the first contexts which were
+	 * added will be run first. On the other hand, if anyone adds a new callback
+	 * for an existing context it will supplant the previously-set callback for
+	 * that context because we reassigned that map key's value.
+	 *
+	 * In the case that a callback adds a new callback to its own context then
+	 * the callback it adds will appear at the end of the iteration and will be
+	 * run only after all other existing contexts have finished executing.
+	 *
 	 * @param {IdleDeadline|number} deadline Idle callback deadline object, or
 	 *                                       animation frame timestamp.
 	 */
 	const runWaitingList = ( deadline ) => {
-		const hasTimeRemaining =
-			typeof deadline === 'number'
-				? () => false
-				: () => deadline.timeRemaining() > 0;
-
-		do {
-			if ( waitingList.length === 0 ) {
-				isRunning = false;
-				return;
+		for ( const [ nextElement, callback ] of waitingList ) {
+			waitingList.delete( nextElement );
```

Comment:
Is it ok to delete an element from a map while we continue iterating. We won't be skipping elements randomly or something? I guess not since the tests look ok, but that's the only weird thing here.